ROX-17013: Implement test in sensor integration tests for testing disconnect/offline behavior

fredrb opened the pull request:
[WIP]

## Description

TODO

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))


## Testing Performed

fredrb commented on sensor/tests/helper/helper.go:
```diff
@@ -538,21 +590,35 @@ func startSensorAndFakeCentral(env *envconf.Config, config CentralConfig) (*cent
+			log.Printf("Sensor stopped")
+		}
+
+		sensorStopped <- true
```
Use signal?

fredrb commented on sensor/tests/helper/helper.go:
```diff
@@ -513,11 +564,11 @@ type CentralConfig struct {
 }
 
-func startSensorAndFakeCentral(env *envconf.Config, config CentralConfig) (*centralDebug.FakeService, func(), func()) {
-	utils.CrashOnError(os.Setenv("ROX_MTLS_CERT_FILE", "../../../../tools/local-sensor/certs/cert.pem"))
```
Fix relative paths

vikin91 commented on sensor/tests/helper/helper.go:
```diff
@@ -1,19 +1,22 @@
-package resource
```
Did you have maybe some other ideas for that name on your shortlist? 😅

fredrb replied:
Unfortunately no. But this whole file is pending a refactor. @lvalerom were talking about breaking this down into smaller components of the test harness. Like managing sensor instance, interacting with the clusters and asserting eventual resources sent to Central are three examples of responsibilities that are bundled in the same file. 

I'd defer choosing a good name for this package to when we break it down in smaller pieces.

lvalerom reviewed (changes requested):
There is a race between `CloseSend` and `SendMsg` in the gRPC stream. I'm not sure if it's something that happens in the reconnect or if this is something that was racy before but it's failing consistently for me with `go test -race -count=1 github.com/stackrox/rox/sensor/tests/connection`.

```
==================
WARNING: DATA RACE
Write at 0x00c001288a90 by goroutine 350:
  google.golang.org/grpc.(*clientStream).CloseSend()
      /Users/lvalerom/go/pkg/mod/google.golang.org/grpc@v1.55.0/stream.go:950 +0x71
  github.com/stackrox/rox/generated/internalapi/central.(*sensorServiceCommunicateClient).CloseSend()
      <autogenerated>:1 +0x49
  github.com/stackrox/rox/sensor/common/sensor.(*centralCommunicationImpl).sendEvents.func2()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_communication_impl.go:146 +0x43
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:476 +0x32
  github.com/stackrox/rox/sensor/common/sensor.(*centralCommunicationImpl).Start.func1()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_communication_impl.go:40 +0x125

Previous read at 0x00c001288a90 by goroutine 476:
  google.golang.org/grpc.(*clientStream).SendMsg()
      /Users/lvalerom/go/pkg/mod/google.golang.org/grpc@v1.55.0/stream.go:866 +0xea
  github.com/stackrox/rox/generated/internalapi/central.(*sensorServiceCommunicateClient).Send()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/generated/internalapi/central/sensor_iservice.pb.go:1647 +0x54
  github.com/stackrox/rox/sensor/common/metrics.countingMessageStream.Send()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/metrics/counting_events.go:35 +0xb3
  github.com/stackrox/rox/sensor/common/metrics.(*countingMessageStream).Send()
      <autogenerated>:1 +0x71
  github.com/stackrox/rox/sensor/common/metrics.timingMessageStream.Send()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/metrics/timing_events.go:16 +0xeb
  github.com/stackrox/rox/sensor/common/metrics.(*timingMessageStream).Send()
      <autogenerated>:1 +0x71
  github.com/stackrox/rox/sensor/common/deduper.(*deduper).Send()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/deduper/deduper.go:79 +0x7e1
  github.com/stackrox/rox/sensor/common/metrics.countingMessageStream.Send()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/metrics/counting_events.go:35 +0xb3
  github.com/stackrox/rox/sensor/common/metrics.(*countingMessageStream).Send()
      <autogenerated>:1 +0x71
  github.com/stackrox/rox/sensor/common/metrics.timingMessageStream.Send()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/metrics/timing_events.go:16 +0xeb
  github.com/stackrox/rox/sensor/common/metrics.(*timingMessageStream).Send()
      <autogenerated>:1 +0x71
  github.com/stackrox/rox/sensor/common/sensor.(*centralSenderImpl).send()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_sender_impl.go:93 +0x8c4
  github.com/stackrox/rox/sensor/common/sensor.(*centralSenderImpl).Start.func1()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_sender_impl.go:19 +0x8b

Goroutine 350 (running) created at:
  github.com/stackrox/rox/sensor/common/sensor.(*centralCommunicationImpl).Start()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_communication_impl.go:40 +0x3d6
  github.com/stackrox/rox/sensor/common/sensor.(*Sensor).communicationWithCentralWithRetries.func1()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/sensor.go:351 +0x507
  github.com/cenkalti/backoff/v3.RetryNotifyWithTimer()
      /Users/lvalerom/go/pkg/mod/github.com/cenkalti/backoff/v3@v3.2.2/retry.go:52 +0x1a7
  github.com/cenkalti/backoff/v3.RetryNotify()
      /Users/lvalerom/go/pkg/mod/github.com/cenkalti/backoff/v3@v3.2.2/retry.go:31 +0x127
  github.com/stackrox/rox/sensor/common/sensor.(*Sensor).communicationWithCentralWithRetries()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/sensor.go:335 +0xd3
  github.com/stackrox/rox/sensor/common/sensor.(*Sensor).Start.func3()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/sensor.go:231 +0x47

Goroutine 476 (running) created at:
  github.com/stackrox/rox/sensor/common/sensor.(*centralSenderImpl).Start()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_sender_impl.go:19 +0x135
  github.com/stackrox/rox/sensor/common/sensor.(*centralCommunicationImpl).sendEvents()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_communication_impl.go:158 +0xf7b
  github.com/stackrox/rox/sensor/common/sensor.(*centralCommunicationImpl).Start.func1()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_communication_impl.go:40 +0x125
==================
```

vikin91 commented:
I guess you still need to fix [the data race](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/stackrox_stackrox/6237/pull-ci-stackrox-stackrox-master-gke-nongroovy-e2e-tests/1665619908144140288) before merging.

vikin91 commented on sensor/tests/helper/helper.go:
```diff
@@ -232,6 +255,48 @@ func (c *TestContext) createTestNs(ctx context.Context, name string) (*v1.Namesp
+// StartFakeGRPC will start a gRPC server to act as Central.
+func (c *TestContext) StartFakeGRPC() {
+	fakeCentral := centralDebug.MakeFakeCentralWithInitialMessages(
+		message.SensorHello("00000000-0000-4000-A000-000000000000"),
```
Oh, if this is a magic number, then maybe we should put it into a var and write a comment?

fredrb replied:
Done

fredrb commented:
@lvalerom: nice catch, thanks for checking this! Seems like the race was already there. We were closing the stream (with `CloseSend()`) on a `defer` function, before stopping the loop that reads/sends into the stream. Essentially, both the defer function and the stream loop tried to use the same stream pointer to close and send at the same time. 

I fixed this by putting some order to the cleaning up of the streams. I moved the teardown of the gRPC stream to the same defered function that sends the stop signals, and also added a `WaitGroup` to make sure that we only close the stream once the subcomponents reported that they finished. 

Still haven't looked at the other comments. But let me know what do you think about the race fix.

fredrb commented on sensor/tests/helper/helper.go:
```diff
@@ -232,6 +255,48 @@ func (c *TestContext) createTestNs(ctx context.Context, name string) (*v1.Namesp
+// StopCentralGRPC will attempt to stop fake central. If it was already stopped, nothing happens
+func (c *TestContext) StopCentralGRPC() {
+	if c.centralStopped.CompareAndSwap(false, true) {
+		c.fakeCentral.Stop()
+		c.fakeCentral.ServerPointer.Stop()
+		messagesBeforeStopping := c.fakeCentral.GetAllMessages()
```
Why? IMO it makes more sense to first stop (i.e. make sure we are not receiving any more messages) and then check what are the received message _after_ the gRPC stopped.

lvalerom replied:
That makes sense. I don't know what I was thinking 😅

fredrb commented on sensor/tests/helper/helper.go:
```diff
@@ -218,6 +236,11 @@ func (c *TestContext) deleteNs(ctx context.Context, name string) error {
 	return nil
 }
 
+// SensorStopped checks if sensor under test stopped.
```
The reference here is the [xUnit](http://xunitpatterns.com/SUT.html) "code under test" or "system under test" often used in testing literature. Not sure if this suggestion makes sense. I'm happy to change it to something else if the wording is confusing though.

lvalerom replied:
I'm not sure what it is but there's something that doesn't read well to me. I'm ok with leaving it as it is.

fredrb commented:
/test gke-nongroovy-e2e-tests

vikin91 commented:
Race fix looks good! Still ✅

fredrb commented on sensor/common/sensor/central_communication.go:
```diff
@@ -17,10 +18,12 @@ type CentralCommunication interface {
-		receiver:   NewCentralReceiver(components...),
-		sender:     NewCentralSender(components...),
-		components: components,
+		allFinished: &finished,
```
But I need to pass `finished` to the two constructors `NewCentralReceiver` and `NewCentralSender` below this line. Not sure this helps 🤔

vikin91 replied:
Oh sorry, I missed that. Ignore my suggestion then
